src/main/java/frc/trigon/robot/poseestimation/poseestimator/PoseEstimator.java

```diff
@@ -113,41 +130,145 @@ private void logTargetPath() {
+    private boolean isVisionObservationTooOld(double timestamp) {
+        try {
+            final double oldestEstimatedRobotPoseTimestamp = previousOdometryPoses.getInternalBuffer().lastKey() - PoseEstimatorConstants.POSE_BUFFER_SIZE_SECONDS;
+            if (oldestEstimatedRobotPoseTimestamp > timestamp)
+                return true;
+        } catch (NoSuchElementException e) {
+            return true;
         }
+        return false;
+    }
+
+    /**
+     * Gets the estimated pose of the robot at the target timestamp.
+     *
+     * @param timestamp the target timestamp
+     * @return the robot's estimated pose at the timestamp
+     */
+    public Pose2d getPoseSample(double timestamp) {
+        final Pose2d samplePose = previousOdometryPoses.getSample(timestamp).orElse(new Pose2d());
+        final Transform2d odometryPoseToSamplePoseTransform = new Transform2d(odometryPose, samplePose);
+
+        return estimatedPose.plus(odometryPoseToSamplePoseTransform);
+    }
+
+    private Transform2d calculatePoseStandardDeviations(Pose2d estimatedPoseAtObservationTime, PoseEstimatorConstants.StandardDeviations cameraStandardDeviations) {
+        final Transform2d poseEstimateAtObservationTimeToObservationPose = new Transform2d(estimatedPoseAtObservationTime, estimatedPose);
+        final PoseEstimatorConstants.StandardDeviations estimatedPoseStandardDeviations = combineOdometryAndVisionStandardDeviations(cameraStandardDeviations);
+        return scaleTransformFromStandardDeviations(poseEstimateAtObservationTimeToObservationPose, estimatedPoseStandardDeviations);
+    }
+
+    private PoseEstimatorConstants.StandardDeviations combineOdometryAndVisionStandardDeviations(PoseEstimatorConstants.StandardDeviations observationStandardDeviation) {
+        final PoseEstimatorConstants.StandardDeviations odometryStandardDeviations = PoseEstimatorConstants.ODOMETRY_STANDARD_DEVIATIONS;
+
+        return new PoseEstimatorConstants.StandardDeviations(
+                combineOdometryAndVisionStandardDeviation(odometryStandardDeviations.translation(), observationStandardDeviation.translation()),
+                combineOdometryAndVisionStandardDeviation(odometryStandardDeviations.theta(), observationStandardDeviation.theta())
+        );
+    }
+
+    /**
+     * Combines a standard deviation value of the odometry and of the vision result to get a new standard deviation.
+     *
+     * @param odometryStandardDeviation a standard deviation value of the estimated odometry pose
+     * @param visionStandardDeviation   a standard deviation value of the estimated vision pose
+     * @return the combined standard deviation
+     */
+    private double combineOdometryAndVisionStandardDeviation(double odometryStandardDeviation, double visionStandardDeviation) {
+        return odometryStandardDeviation / (odometryStandardDeviation + Math.sqrt(odometryStandardDeviation * visionStandardDeviation));
+    }
```

Comment:
I would make this a function of the record (with a jdoc

Reply:
Is this a standard thing for combining standard deviations or just for odom and vis?

Reply:
idk
but it could be lile Translation2d.plus(other), you could do a static function that accepts two but it's more nice the other way